File: ruby/07.rb
#!/usr/bin/env ruby

@crabs = File.open("../inputs/input07.txt", "r").each_line.first.split(",").map(&:to_i)

"""
Find fuel amount (r) needed to minimise sum of movements to align all crab positions.
"""
def part1
    lowest_sum = 9999999

    (@crabs.min..@crabs.max).map do |r|
        deltas = @crabs.map{ |c| (c - r).abs }
        sum = deltas.reduce(:+)
        if sum <= lowest_sum
            lowest_sum = sum
        else
            break lowest_sum
        end
    end
end

"""
Find fuel amount (r) needed to minimise sum of movements to align all crab positions.
Larger movements cost more, e.g. moving 5 costs 1+2+3+4+5
"""
def part2
    lowest_sum = 9999999999

    (@crabs.min..@crabs.max).map do |r|
        deltas = @crabs.map{ |c| (c - r).abs }
        sum = deltas.map{ |d|
            d > 0 ? (1..d).reduce(:+) : 0
        }.reduce(:+)
        if sum <= lowest_sum
            lowest_sum = sum
        else
            break lowest_sum
        end
    end
end

p "Part 1: #{part1}" # P1: 328187
p "Part 2: #{part2}" # P2: 91257582

File: elixir/07.exs
defmodule Day07 do
  defp read_input do
    File.read!(Path.expand("../inputs/input07.txt"))
    |> String.trim
    |> String.split(",")
    |> Enum.map(&String.to_integer/1)
  end

  @doc """
  Find fuel amount (r) needed to minimise sum of movements to align all crab positions.
  """
  def part1 do
    crabs = read_input()

    (Enum.min(crabs)..Enum.max(crabs))
    |> Enum.map(fn r ->
        crabs
        |> Enum.map(fn c -> abs(c - r) end)
        |> Enum.sum
    end)
    |> Enum.min
    |> IO.inspect
  end

  @doc """
  Find fuel amount (r) needed to minimise sum of movements to align all crab positions.
  Larger movements cost more, e.g. moving 5 costs 1+2+3+4+5
  """
  def part2 do
    crabs = read_input()

    (Enum.min(crabs)..Enum.max(crabs))
    |> Enum.map(fn r ->
        crabs
        |> Enum.map(fn c -> abs(c - r) end)
        |> Enum.map(fn d ->
          cond do
            d > 0 -> Enum.sum(1..d)
            true -> 0
          end
        end)
        |> Enum.sum
    end)
    |> Enum.min
    |> IO.inspect
  end
end

# P1: 328187
# P2: 91257582
